### PR TITLE
Updated Signelr/Websockets

### PR DESCRIPTION
**Pull request by YAS-CODE**
Updated task https://github.com/webRunes/WRIO-InternetOS/issues/1891

### CODE REVIEW

File: src/base/components/DeviceProfile.js [hunk 1]
```diff
@@ -43,10 +102,16 @@ class DeviveProfileTab extends React.Component {
...
-        this.setState({ devicestatus: "on"});
+        if(data.status){
+          if(this.state.devicestatus === "on" )
+            this.setState({ devicestatus: "off"});
+          else
+          this.setState({ devicestatus: "on"});
+        }
+        else{
+          this.setState({ devicestatus: "on"});
+          toast.error("Device seems offline, Please start Experiment!!!");
```
Comment (AlexeyAnshakov) [inline src/base/components/DeviceProfile.js]: The device is offline. Please re-start the experiment

File: src/base/components/FeedList.js [hunk 2]
```diff
@@ -25,8 +32,83 @@ class FeedListPage extends React.Component {
...
+          console.log("ReceiveMessage:["+topic+"] "+ jdata);
+          
+          if(topic === "/iot/zolertia/replay"){
+            jdata =  NormailizeJSON(jdata);
+            var data = JSON.parse(jdata);
+            if(data.Command == "on" && data.Result == "Success"){
+              toast.success("Zolertia Successfully Turned ON!!!");
+            }
+            else if(data.Command == "off" && data.Result == "Success"){
+              toast.success("Zolertia Successfully Turned OFF!!!");
```
Comment (AlexeyAnshakov) [inline src/base/components/FeedList.js]: The device has been successfully turned off

File: src/base/components/FeedList.js [hunk 3]
```diff
@@ -25,8 +32,83 @@ class FeedListPage extends React.Component {
...
+            }
+            else if(data.Command == "SetTX" && data.Result == "Success"){
+              toast.success("Successfully Set "+data.Value+" to TX Level!!!");
```
Comment (AlexeyAnshakov) [inline src/base/components/FeedList.js]: "TX Level" + data.Value+"has been set"

File: src/base/components/FeedList.js [hunk 4]
```diff
@@ -25,8 +32,83 @@ class FeedListPage extends React.Component {
...
+            }
+            else if (data.Command == "SetCh" && data.Result ==  "Success"){
+              toast.success("Successfully Set "+data.Value+" Channel!!!");
```
Comment (AlexeyAnshakov) [inline src/base/components/FeedList.js]: "Channel" + data.Value+"has been set"

File: src/base/components/FeedList.js [hunk 5]
```diff
@@ -25,8 +32,83 @@ class FeedListPage extends React.Component {
...
+            }
+            else if(data.Commnad == "SetTX" && data.Result == "Failed"){
+              toast.error("Failed to Set "+data.Value+" to TX Level!!!");
```
Comment (AlexeyAnshakov) [inline src/base/components/FeedList.js]: "Something goes wrong. TX Level remains the same"

File: src/base/components/FeedList.js [hunk 6]
```diff
@@ -25,8 +32,83 @@ class FeedListPage extends React.Component {
...
+            }
+            else if (data.Command == "SetCh" && data.Result ==  "Failed"){
+              toast.error("Failed to Set "+data.Value+" Channel!!!");
```
Comment (AlexeyAnshakov) [inline src/base/components/FeedList.js]: "Something goes wrong. Channel remains the same"

File: src/base/components/FeedList.js [hunk 7]
```diff
@@ -25,8 +32,83 @@ class FeedListPage extends React.Component {
...
+          }
+          else if(topic === "/iot/sensor/data" ){
+            jdata =  NormailizeJSON(jdata);
+            var data = JSON.parse(jdata);
+            this.setState({ temperature: (parseFloat(data.Temperature)/1000)});
+            this.setState({ battery: (parseFloat(data.BatteryVal)/1000)});
+            this.setState({ devicestatus: "off"});
+          }
+          else if(topic === "/iot/sensor/boot" && jdata.includes("BOOTUP")){
+            toast.success("Zolertia Sucessfully Bootup now!!!");
```
Comment (AlexeyAnshakov) [inline src/base/components/FeedList.js]: The device is online

File: src/base/components/FeedList.js [hunk 8]
```diff
@@ -25,8 +32,83 @@ class FeedListPage extends React.Component {
...
+          }
+          else if(topic === "/iot/sensor/error"){
+            if(jdata.includes("SERIALACCESSERROR")){
+              toast.error("Zolertia does not FOUND!!!");
```
Comment (AlexeyAnshakov) [inline src/base/components/FeedList.js]: The device can't be found

File: src/base/components/FeedList.js [hunk 9]
```diff
@@ -25,8 +32,83 @@ class FeedListPage extends React.Component {
...
+              this.setState({ devicestatus: "on"});
+            }            
+            if(jdata.includes("SERIALREADWRITEERROR")){
+              toast.error("Zolertia read write error FOUND!!!");
```
Comment (AlexeyAnshakov) [inline src/base/components/FeedList.js:82]: Not sure what does this message mean, please clerify

File: src/base/components/FeedList.js [hunk 10]
```diff
@@ -25,8 +32,83 @@ class FeedListPage extends React.Component {
...
+  .build();
+      
+    connection.on("ReceiveMessage", (topic ,jdata) => {
+          console.log("ReceiveMessage:["+topic+"] "+ jdata);
+          
+          if(topic === "/iot/zolertia/replay"){
+            jdata =  NormailizeJSON(jdata);
+            var data = JSON.parse(jdata);
+            if(data.Command == "on" && data.Result == "Success"){
+              toast.success("Zolertia Successfully Turned ON!!!");
```
Comment (AlexeyAnshakov) [inline src/base/components/FeedList.js]: The device has been successfully turned on

File: src/base/components/FeedList.js [hunk 11]
```diff
@@ -25,8 +32,83 @@ class FeedListPage extends React.Component {
...
+        return console.error("signalr error: "+err.toString());
+    })
+
+    try {
+      this.interval = setInterval(async () => {
+        const res = await fetch(gconfig.gatewaysServiceUrl+"/api/Device/IsDeviceConnected");
+        const devstatus = await res.json();
+        if(devstatus!= null && devstatus.status == false){
+          this.setState({ devicestatus: "on"});        
+          toast.error("Please start your Experiment, Thanks!!!");
```
Comment (AlexeyAnshakov) [inline src/base/components/FeedList.js]: Please re-start the experiment

File: src/base/components/FeedList.js [hunk 12]
```diff
@@ -114,10 +196,16 @@ class FeedListPage extends React.Component {
...
-      this.setState({ devicestatus: "on"});
+      if(data.status){
+        if(this.state.devicestatus === "on" )
+          this.setState({ devicestatus: "off"});
+        else
+        this.setState({ devicestatus: "on"});
+      }
+      else{
+        this.setState({ devicestatus: "on"});
+        toast.error("Device seems offline, Please start Experiment!!!");
```
Comment (AlexeyAnshakov) [inline src/base/components/FeedList.js]: The device is offline. Please re-start the experiment

**Review by AlexeyAnshakov — CHANGES_REQUESTED** [review]

File: src/base/components/FeedList.js (hunk 9 above)
Comment (YAS-CODE) [reply src/base/components/FeedList.js:82]: During read/write from Device of if any error comes up, then trigger this event, its very rare

File: src/base/components/FeedList.js (hunk 12 above)
Comment (YAS-CODE) [reply src/base/components/FeedList.js]: Done

File: src/base/components/FeedList.js (hunk 11 above)
Comment (YAS-CODE) [reply src/base/components/FeedList.js]: Done

File: src/base/components/FeedList.js (hunk 10 above)
Comment (YAS-CODE) [reply src/base/components/FeedList.js]: Done

File: src/base/components/FeedList.js (hunk 8 above)
Comment (YAS-CODE) [reply src/base/components/FeedList.js]: Done

File: src/base/components/FeedList.js (hunk 7 above)
Comment (YAS-CODE) [reply src/base/components/FeedList.js]: Done

File: src/base/components/FeedList.js (hunk 6 above)
Comment (YAS-CODE) [reply src/base/components/FeedList.js]: Done

File: src/base/components/FeedList.js (hunk 5 above)
Comment (YAS-CODE) [reply src/base/components/FeedList.js]: Done

File: src/base/components/FeedList.js (hunk 4 above)
Comment (YAS-CODE) [reply src/base/components/FeedList.js]: Done

File: src/base/components/FeedList.js (hunk 3 above)
Comment (YAS-CODE) [reply src/base/components/FeedList.js]: Done

File: src/base/components/FeedList.js (hunk 2 above)
Comment (YAS-CODE) [reply src/base/components/FeedList.js]: Done

File: src/base/components/DeviceProfile.js (hunk 1 above)
Comment (YAS-CODE) [reply src/base/components/DeviceProfile.js]: Done

**Pull request merged**